2/2 DNF kickstart support

jkonecny12 opened the pull request:
Add kickstart support for DNF module.

In this PR I'm disabling `%packages` which will be solved by the main process for the time being. Right now, we have to add support for the base repositories to support subscription spoke.

Will be implemented:
- [x] KS support for HDD
- [x] KS support for NFS
- [x] KS support for URL

Part of this PR was split to https://github.com/rhinstaller/anaconda/pull/2484 to make this smaller.

pep8speaks commented:
Hello @jkonecny12! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:





* In the file [`pyanaconda/modules/payloads/payload/dnf/dnf.py`](https://github.com/rhinstaller/anaconda/blob/fd669bd85ac27f00b59d02401fd65f99569aecdd/pyanaconda/modules/payloads/payload/dnf/dnf.py):

> [Line 63:100](https://github.com/rhinstaller/anaconda/blob/fd669bd85ac27f00b59d02401fd65f99569aecdd/pyanaconda/modules/payloads/payload/dnf/dnf.py#L63): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (101 > 99 characters)

rvykydal commented on pyanaconda/modules/payloads/payload/dnf/dnf.py:
```diff
@@ -43,8 +43,13 @@ def type(self):
+            SourceType.HMC,
+            SourceType.NFS,
+            SourceType.REPO_FILES,
+            SourceType.URL]
```
I'd move the ']' to the next line.

jkonecny12 replied:
Fixed.

rvykydal commented on pyanaconda/modules/payloads/source/factory.py:
```diff
@@ -57,3 +57,25 @@ def create_source(source_type: SourceType):
+        """Get a payload source type for the given kickstart data.
+
+        :param data: a kickstart data
+        :return: a payload type
```
"a payload source type" ?

jkonecny12 replied:
The solution changed during your review :).

jkonecny12 commented:
jenkins, test this please

poncovka commented on pyanaconda/modules/payloads/payload/payload_base.py:
```diff
@@ -135,7 +135,7 @@ def set_sources(self, sources):
 
         self._sources = sources
-        log.debug("New sources %s was added.", sources)
+        log.debug("New sources %s were added.", sources)
```
It would be good to implement the `__str__` method of `PayloadSourceBase` to get some useful info here.

jkonecny12 replied:
Fixed.

poncovka commented on pyanaconda/modules/payloads/payload/dnf/dnf.py:
```diff
@@ -43,16 +47,40 @@ def type(self):
     def process_kickstart(self, data):
         """Process the kickstart data."""
-        pass
+        source_types = SourceFactory.get_type_for_kickstart(data)
```
I don't think this is working. Let's move the method `get_type_for_kickstart` here, because it doesn't really make sense to have it in the factory since only the DNF module is interested in the RPM kickstart sources. It will help with issues regarding mixing sources of different payloads or supporting multiple sources that would come up later.

jkonecny12 replied:
I had the same idea, however, I wanted to keep the same solution we have even now.

I agree that this should be changed. It's unnecessarily complicated solution.

jkonecny12 replied:
I've tried a new solution. See the changed version.

poncovka commented on pyanaconda/modules/payloads/payload/dnf/dnf.py:
```diff
@@ -43,16 +47,40 @@ def type(self):
+            types_strs = map(lambda x: "\"{}\"".format(x.ks_data.currentCmd), source_types)
+            msg = "Only one of {} payload sources can be used!".format(", ".join(types_strs))
+            lineno = source_types[-1].ks_data.lineno
+            raise KickstartParseError(msg, lineno=lineno)
```
This check should be somehow included in the function `get_type_for_kickstart` or moved to a new function or to `parse` methods of the kickstart commands in `pyanaconda/modules/payloads/kickstart.py` or to pykickstart. This solution is scary.

jkonecny12 replied:
The same as above. I've tried a different approach.

poncovka commented on pyanaconda/modules/payloads/payload/dnf/dnf.py:
```diff
@@ -43,16 +47,40 @@ def type(self):
+        self.add_source(source_module)
+
+        # FIXME: This is a temporary workaround.
+        PayloadSourceContainer.to_object_path(source_module)
```
This shouldn't be necessary if the `add_source` method changes a DBus property.

jkonecny12 replied:
I don't follow. How it should change the DBus property from the implementation class?

jkonecny12 replied:
Solved.

poncovka commented on tests/nosetests/pyanaconda_tests/module_payload_shared.py:
```diff
@@ -100,51 +100,58 @@ def prepare_source(source_type, state=SourceState.READY):
-            self._test.assertTrue(self.payload_interface.HasSource())
-
-        return ret
+            self._test.assertFalse(self.payload_interface.HasSource())
```
The DBus method `HasSource` will be removed.

jkonecny12 replied:
Lets do it with a different PR which will remove the `HasSource()` API.
Also why it should go away?

poncovka replied:
It is redundant. You can get the same information by checking the `Sources` property.

jkonecny12 replied:
OK, I'll remove it to safe lines later.

poncovka replied:
There is an issue for that in Jira now.

jkonecny12 commented on pyanaconda/modules/payloads/payload/factory.py:
```diff
@@ -57,7 +57,13 @@ def get_type_for_kickstart(cls, data):
+
+        for method in ["cdrom", "harddrive", "hmc", "nfs", "url"]:
+            method = getattr(data, method)
+            if method.seen:
```
Seems better to read to me but no objections to change it.

jkonecny12 replied:
Fixed.

jkonecny12 commented on pyanaconda/modules/payloads/payloads.py:
```diff
@@ -106,9 +106,9 @@ def process_kickstart(self, data):
+        # FIXME: Process packages in the DNF module as structure. Disabled for now.
         # The packages module should be replaces with a DBus structure.
-        self._packages.process_kickstart(data)
+        # self._packages.process_kickstart(data)
```
Good point with RHEL-8. The reason why I commented it instead of removing is that removed code is hard to find in commits history.

jkonecny12 replied:
Fixed.

jkonecny12 commented on pyanaconda/modules/payloads/payload/payload_base.py:
```diff
@@ -146,6 +146,15 @@ def has_source(self):
+
+        :param source: Source we want to add.
+        """
+        sources = self.sources.copy()
```
Fixed.

jkonecny12 commented:
**UPDATED:**
* Rebased.
* Updated tests to use new naming.
* Implemented some suggestions from @poncovka .
* Implemented suggestion from @rvykydal .
* Add support for harddrive source (excluding --biospart because it does not really work with the current implementation).
* Add new solution for DNF source creation. Let's discuss this approach. If it will get approval I'll replace old commits by this new solution. Right now it's a new (last) commit.

jkonecny12 commented:
**UPDATED:**
* Remove unused imports and fix one more from @poncovka

poncovka commented on pyanaconda/modules/payloads/source/harddrive/harddrive.py:
```diff
@@ -62,6 +64,10 @@ def get_state(self):
     def process_kickstart(self, data):
         """Process the kickstart data."""
+        if data.harddrive.biospart:
+            raise KickstartParseError("biospart support is not implemented",
```
Please, translate the string and provide more information. It is not clear that you are talking about the `harddrive` command and its `--biospart` option.

jkonecny12 commented on tests/nosetests/pyanaconda_tests/module_payload_dnf_test.py:
```diff
@@ -21,20 +21,31 @@
+        """Test DNF supported sources API."""
+        self.assertEqual(
+            [SourceType.CDROM.value,
+             SourceType.HDD.value,
```
These constants were not available before. I'll change it now.

jkonecny12 replied:
Fixed.

jkonecny12 commented on pyanaconda/modules/payloads/payload/dnf/kickstart_factory.py:
```diff
@@ -0,0 +1,84 @@
+        if self._ks_data.url.seen:
+            self._set_and_set_source(self._ks_data.url, SourceType.URL)
+
+    def _set_and_set_source(self, source_ks_data, source_type):
```
ups typo, but it will change so or so :)

jkonecny12 commented on pyanaconda/modules/payloads/payload/factory.py:
```diff
@@ -57,7 +57,11 @@ def get_type_for_kickstart(cls, data):
+        data.harddrive.seen or \
+        data.hmc.seen or \
+        data.nfs.seen or \
+        data.url.seen:
```
Fixed.

jkonecny12 commented on pyanaconda/modules/payloads/payloads.py:
```diff
@@ -129,7 +125,7 @@ def generate_temporary_kickstart(self):
         # generate packages section only for DNF module
         if isinstance(self.payload, DNFModule):
-            self._packages.setup_kickstart(data)
+            pass
```
Fixed.

jkonecny12 commented:
**UPDATED:**
* Fix tests
* Rename `KickstartFactory` to `KickstartConverter` and change implementation.

poncovka commented on pyanaconda/modules/payloads/payload/dnf/kickstart_converter.py:
```diff
@@ -0,0 +1,83 @@
+        :return: instance of source object
+        :raise: ParseKickstartError if kickstart data are not correct
+        """
+        source_type = cls._process_kickstart(ks_data)
```
This implementation is pretty much the same as `get_type_for_kickstart` from `PayloadFactory`. Because of that, I suggest to move the method `_process_kickstart` to `SourceFactory`, rename it to `get_rpm_type_for_kickstart` and move the content of the method `convert` to the method `process_kickstart` of the class `DNFModule`. It shouldn't cause any new issues and we will move closer to the "common practice".

jkonecny12 replied:
I don't think it is a good way to go but yeah we've already went to this rabbit hole and it's too late to change it here.
We should definitely investigate the possibility to move kickstart processing to a separate space for the module. It's a shame that most of the code in Payload have to deal with KS in some way.

But I agree that we should stay consistent with an existing solution and change this later.

jkonecny12 commented:
**UPDATED:**
* Added NFS, URL kickstart support.

This PR is finished by this and ready for review.

jkonecny12 commented:
**UPDATED:**
* Implemented note from @poncovka .
* Remove source exclusivity. We will behave as pykickstart which is just taking first based on its internal logic. We still want to make this explicit but we should do that in Pykickstart not in the Anaconda. This simplified the code a lot.
* Merge rest of `KickstartConverter` to `DNFPayload` and `SourceFactory`.
* In the process I've squashed three commits creating the ks data processing to the first commit.
* Add new commit to move HMC on the bottom of the list to make it consistent with pykickstart's `method`.

poncovka commented on pyanaconda/modules/payloads/payload/dnf/dnf.py:
```diff
@@ -54,11 +56,21 @@ def supported_source_types(self):
+        source_type = SourceFactory.get_rpm_type_for_kickstart(data)
+
+        if source_type is None:
+            return None
```
The method `process_kickstart` doesn't return anything.

jkonecny12 replied:
Good point. Thanks.

jkonecny12 commented:
**UPDATED:**
* Remove leftover import.

poncovka commented on pyanaconda/modules/payloads/payload/dnf/dnf.py:
```diff
@@ -54,11 +55,21 @@ def supported_source_types(self):
+
+        source = SourceFactory.create_source(source_type)
+
+        if source is not None:
```
The method `create_source` will raise an exception if the source is unknown. Anyway, it is a bug if the factory doesn't know how to create a source returned by `get_rpm_type_for_kickstart`, so I don't think we need to handle it.

jkonecny12 replied:
ACK.

poncovka commented on pyanaconda/modules/payloads/source/factory.py:
```diff
@@ -57,3 +57,18 @@ def create_source(source_type: SourceType):
         raise ValueError("Unknown source type: {}".format(source_type))
+
+    @staticmethod
+    def get_rpm_type_for_kickstart(ks_data):
```
Please, add a short docstring.

jkonecny12 replied:
Hmm I missed docstring here. Thanks!

poncovka commented on pyanaconda/modules/payloads/source/nfs/nfs.py:
```diff
@@ -46,6 +47,20 @@ def for_publication(self):
+
+    def setup_kickstart(self, data):
+        """Setup the kickstart data."""
+        (opts, host, path) = parse_nfs_url(self.url)
```
This is out of scope of this PR, but shouldn't we replace the URL with a DBus structure?

jkonecny12 replied:
I don't think so. It's easier to parse and merge the url so there is no need to complicate an API here. From my PoV.
Also we're doing the same with proxy strings.

poncovka commented on pyanaconda/modules/payloads/source/url/url.py:
```diff
@@ -67,6 +68,46 @@ def for_publication(self):
+        repo_data.ssl_verification_enabled = not data.url.noverifyssl
+        repo_data.ssl_configuration.ca_cert_path = data.url.sslcacert
+        repo_data.ssl_configuration.client_cert_path = data.url.sslclientcert
+        repo_data.ssl_configuration.client_key_path = data.url.sslclientkey
```
Are you sure this cannot be None?

jkonecny12 replied:
I don't think it will be. In general there should always be `False`/`True`. I would leave it like this and we can fix it later if there will be an error later.

jkonecny12 replied:
Fixed I thought we are talking just about the first line.

jkonecny12 commented:
**UPDATED:**
* Implemented notes from @poncovka .
